File: opencl/01_tutorial/main.cpp
#include "cll.hpp"

class myAlgo: public CL
{
public:
    // function to use for vector add
    cl::make_kernel<cl::Buffer, cl::Buffer, cl::Buffer> func_vadd;
    myAlgo(std::string clfile):CL(clfile),
                               func_vadd(clprogram, "vadd"){
    }
    virtual void runAlgo();
    virtual ~myAlgo(){}
};

void myAlgo::runAlgo(){
    int length = 10;
    // create vectors: 2 inputs and 1 output
    std::vector<float> v_a(length);
    std::vector<float> v_b(length);
    std::vector<float> v_c (length);

    // create buffers used as vectors in opencl
    cl::Buffer d_a;
    cl::Buffer d_b;
    cl::Buffer d_c;

    // set values of the vectors
    for(int i = 0; i < length; i++) {
        v_a[i]  = i;
        v_b[i]  = i;
    }

    // pass the value of the vectors to the buffers
    vectToBuffer(v_a, d_a);
    vectToBuffer(v_b, d_b);

    // create a buffer with the proper lenght for result
    d_c = outBuffer(length);

    // run opencl function
    func_vadd(cl::EnqueueArgs(queue,cl::NDRange(length)),
         d_a, d_b, d_c);

    // pass the result information into a vector
    bufferToVect(d_c, v_c);

    // printing
    std::cout << "c: ";
    for (int i = 0; i < length; i++) {
        std::cout << v_c[i] << ", ";
    }
    std::cout << "\n";

}

int main(int argc, char *argv[])
{
    std::string clfile = "vector_add.cl";
    myAlgo example(clfile);
    example.runAlgo();
    return 0;
}

File: opencl/01_tutorial/main.py
# from Adventures in OpenCL Part1 to PyOpenCL

import pyopencl as cl
import numpy as np

class CL:
    """
    This class have all the configuration of to load and run
    an openCL program
    """
    def __init__(self):
        self.ctx = cl.create_some_context()
        self.queue = cl.CommandQueue(self.ctx)
        self.mf = cl.mem_flags

    def loadProgram(self, filename):
        """
        The file with the opencl program is loaded
        """
        # read in the OpenCL source file as a string
        f = open(filename, 'r')
        fstr = "".join(f.readlines())
        # create the program
        self.program = cl.Program(self.ctx, fstr).build()

    def vectToBuffer(self, vector):
        """ Converts numpy array to opencl array"""
        return cl.Buffer(self.ctx, self.mf.READ_ONLY | self.mf.COPY_HOST_PTR, hostbuf=vector)

    def bufferToVect(self, buff, vect):
        """ Converts opencl array to numpy array"""
        cl.enqueue_read_buffer(self.queue, buff, vect).wait()
        
    def outBuffer(self, nbytes):
        """creates output for opencl """
        return cl.Buffer(self.ctx, self.mf.WRITE_ONLY, nbytes)
    
    
    def runAlgo(self):
        """
        The program is created here
        """
        #initialize client side (CPU) arrays
        a = np.array(range(10), dtype=np.float32)
        b = np.array(range(10), dtype=np.float32)

        #create OpenCL buffers
        a_buf = self.vectToBuffer(a)
        b_buf = self.vectToBuffer(b)
        dest_buf = self.outBuffer(b.nbytes)
        # execute program
        self.program.vadd(self.queue, a.shape, None,
                          a_buf, b_buf, dest_buf)
        c = np.empty_like(a)
        self.bufferToVect(dest_buf, c)

        # print result
        print "a", a
        print "b", b
        print "c", c

if __name__ == "__main__":
    example = CL()
    example.loadProgram("vector_add.cl")
    example.runAlgo()

